query-parser/src/parse.rs

```diff
@@ -172,37 +207,142 @@ impl<I> FindSp<I>
 mod test {
     extern crate combine;
     extern crate edn;
     extern crate mentat_query;
+    extern crate ordered_float;
 
     use self::combine::Parser;
-    use self::mentat_query::{Element, FindSpec, Variable};
+    use self::ordered_float::OrderedFloat;
+    use self::mentat_query::{
+        Element,
+        FindSpec,
+        NonIntegerConstant,
+        Pattern,
+        PatternNonValuePlace,
+        PatternValuePlace,
+        SrcVar,
+        Variable,
+    };
 
     use super::*;
 
     #[test]
+    fn test_pattern_mixed() {
+        let e = edn::PlainSymbol::new("_");
+        let a = edn::NamespacedKeyword::new("foo", "bar");
+        let v = OrderedFloat(99.9);
+        let tx = edn::PlainSymbol::new("?tx");
+        let input = [edn::Value::Vector(
+            vec!(edn::Value::PlainSymbol(e.clone()),
+                 edn::Value::NamespacedKeyword(a.clone()),
+                 edn::Value::Float(v.clone()),
+                 edn::Value::PlainSymbol(tx.clone())))];
+        assert_parses_to!(Where::pattern, input, Pattern {
+            source: None,
+            entity: PatternNonValuePlace::Placeholder,
+            attribute: PatternNonValuePlace::Ident(a),
+            value: PatternValuePlace::Constant(NonIntegerConstant::Float(v)),
+            tx: PatternNonValuePlace::Variable(Variable(tx)),
+        });
+    }
+
+    #[test]
+    fn test_pattern_vars() {
+        let s = edn::PlainSymbol::new("$x");
+        let e = edn::PlainSymbol::new("?e");
+        let a = edn::PlainSymbol::new("?a");
+        let v = edn::PlainSymbol::new("?v");
+        let tx = edn::PlainSymbol::new("?tx");
+        let input = [edn::Value::Vector(
+            vec!(edn::Value::PlainSymbol(s.clone()),
+                 edn::Value::PlainSymbol(e.clone()),
+                 edn::Value::PlainSymbol(a.clone()),
+                 edn::Value::PlainSymbol(v.clone()),
+                 edn::Value::PlainSymbol(tx.clone())))];
+        assert_parses_to!(Where::pattern, input, Pattern {
+            source: Some(SrcVar::NamedSrc("x".to_string())),
+            entity: PatternNonValuePlace::Variable(Variable(e)),
+            attribute: PatternNonValuePlace::Variable(Variable(a)),
+            value: PatternValuePlace::Variable(Variable(v)),
+            tx: PatternNonValuePlace::Variable(Variable(tx)),
+        });
+    }
+
+    #[test]
+    fn test_pattern_reversed_invalid() {
+        let e = edn::PlainSymbol::new("_");
+        let a = edn::NamespacedKeyword::new("foo", "_bar");
+        let v = OrderedFloat(99.9);
+        let tx = edn::PlainSymbol::new("?tx");
+        let input = [edn::Value::Vector(
+            vec!(edn::Value::PlainSymbol(e.clone()),
+                 edn::Value::NamespacedKeyword(a.clone()),
+                 edn::Value::Float(v.clone()),
+                 edn::Value::PlainSymbol(tx.clone())))];
+
+        let mut par = Where::pattern();
+        let result = par.parse(&input[..]);
+        if let Err(_) = result {
```

Comment:
There's a `matches` crate that allows to use `match` as an expression, and then `assert!(matches!(...))`.  Might be worth using.